src/types.ts

```diff
@@ -313,6 +313,7 @@ export interface ExcalidrawProps {
     pointerDownState: PointerDownState,
   ) => void;
   onScrollChange?: (scrollX: number, scrollY: number) => void;
+  defaultTheme?: Theme;
```

Comment:
we should use `THEME.LIGHT` instead as the default theme instead of providing another prop. Also lets make a constant `DEFAULT_APP_THEME` with value `THEME.LIGHT`

Reply:
@ad1992 sorry but I don't see how that would help in this problem.

By **_default theme_**, I meant the theme value that was stored in the local storage from the previous session.

so, essentially what we want to achieve is to know the default theme value before rendering the `App`.  This will allow us to show `<LoadingMessage />` and the default state of the `<App />` to be set to the theme from the local storage if there is any.

So the reason behind introducing the `defaultTheme` prop is to essentially pass theme value from local storage to the `<App/>` component.

```mermaid
graph TD;
    theme(light/dark)--> localStorage;
    localStorage--> Excalidraw_Component;
    Excalidraw_Component--> InitializeApp_Component;
    InitializeApp_Component--> LoadingMessage_Component;
    Excalidraw_Component--> App_Component;
```
> Also lets make a constant `DEFAULT_APP_THEME` with value `THEME.LIGHT`

this won't make any difference as the `constructor` of the `App` component already sets the theme value in the default state to light - returned by the `getDefaultAppState` function. 👇

```ts
constructor(props: AppProps) {
    super(props);
    const defaultAppState = getDefaultAppState();
    const {
      excalidrawRef,
      viewModeEnabled = false,
      zenModeEnabled = false,
      gridModeEnabled = false,
      theme = defaultAppState.theme,
      name = defaultAppState.name,
    } = props;
    this.state = {
      ...defaultAppState,
      theme,
      isLoading: true,
      ...this.getCanvasOffsets(),
      viewModeEnabled,
      zenModeEnabled,
      gridSize: gridModeEnabled ? GRID_SIZE : null,
      name,
      width: window.innerWidth,
      height: window.innerHeight,
      showHyperlinkPopup: false,
      isLibraryMenuDocked: false,
    };
```

### Alternate Approach

Another approach to avoid `defaultTheme` prop from `<ExcalidrawBase />` is to resolve `initialData` and get the theme value from that inside a `useEffect` in `ExcalidrawBase` component and set the result to some state. Then that state can be passed to `InitializeApp` and `App` component as the default theme values.

Reply:
@ad1992 any alternate approach for the above 👆 mentioned problem?

Reply:
If we go this route,
The 'theme' prop will always take precedence over `defaultProp` for the loading screen, so the devs provide whatever theme value will override it.

Reply:
@mabdullahadeel yes lets try with `initialData` to get the initial theme value.
 The `theme` prop is anyways always going to be higher precedence when present over initial data.

Reply:
@ad1992 I have tried this approach 👇

```ts

const [defaultTheme, setDefaultTheme] = useState<Theme>("");

useEffect(() => {
    const resolveInitialData = async () => {
      if (initialData instanceof Promise) {
        const data = await initialData;
        if (data?.appState?.theme) {
          setDefaultTheme(data.appState.theme);
        }
      }
    };
    resolveInitialData();
  }, [initialData]);
```

But this did not solve the problem. The behavior is identical.
Start with a light theme and shift to dark after the app renders.

my guess is that the promise is being resolved after the app renders. Making render dependant on the presence of `defaultTheme` state causes the promise to never resolve inside the `useEffect`. It may be due to `memo` but I don't see that happening since the `defaultTheme` state is being changed.

But overall, I think this idea did not work 😐.  In this situation, the only reasonable thing is to pass the theme as a prop to `ExcalidrawBase` which is how it's done in the current state of this PR.

Reply:
I agree with both of you. We definitely need to fix this issue. But having two props for theme (`props.theme` and `props.defaultTheme`) is not great, even though `defaultTheme` does solve this problem.

What we want is the ability to have the theme controlled (`props.theme`) but also show the UI toggle, _and_ still have the DX ok-ish.

So I suggest we keep the `props.theme` behavior as is, but make the (already existing) `props.UIOptions.canvasOptions.theme` to work in this way: default to `undefined` which means we'll render the theme toggle if `props.theme` is `undefined`. If `false`, never render it. If `true`, always render if regardless of `props.theme`.

We'll also want to factor out the theme persistence from `appState.theme` (`STORAGE_KEYS.LOCAL_STORAGE_APP_STATE`) to `theme` (`STORAGE_KEYS.THEME`). Not strictly necessary, but we want to be able to quickly access the theme without having to import the whole state (`importFromLocalStorage()`). App theme is semantically separate from editor theme anyway.

Reply:
@dwelle 
Thanks, this approach worked!